File: zipf.cc
// Copyright © 2010 Bart Massey
// Illustration of "Zipf's Law"

// This program is released under the terms of the "MIT
// License".  Please see the file COPYING in this
// distribution for license details.

#include <ctype.h>
#include <stdio.h>

#include <list>
#include <map>
#include <string>

using namespace std;

map <string, int> freq_map;
string current_word;
bool in_word = false;

// Count the alphabetic words on the input into the frequency map.
void
count_words()
{
    int    ch;
    bool   in_word = false;
    string current_word;
    do {
        ch = getchar();
        if (isascii(ch) && isalpha(ch)) {
            if (in_word) {
                current_word += tolower(ch);
            } else {
                current_word = string(1, tolower(ch));
                in_word = true;
            }
        } else {
            if (in_word) {
                freq_map[current_word]++;
                in_word = false;
            } else {
                /* do nothing */
            }
        }
    } while (ch != EOF);
}

bool count_compare(pair<string, int> fc1, pair<string, int> fc2) {
    return fc1.second < fc2.second;
}

void show_freqs() {
    /* sort the frequencies */
    map<string, int>::iterator itm;
    list< pair<string, int> > freq_list;
    for (itm = freq_map.begin(); itm != freq_map.end(); itm++)
	freq_list.push_back(*itm);
    freq_list.sort(count_compare);
    /* actually show the calculation */
    int i = freq_list.size();
    list< pair<string, int> >::iterator itp;
    for (itp = freq_list.begin(); itp != freq_list.end(); itp++) {
	int c = itp->second;
	printf("%g %d\n", 1.0 / i--, c);
    }
}

int main ()
{
    count_words();
    show_freqs();
    return 0;
}

File: zipf.hs
-----------------------------------------------------------------------------
-- |
-- Module      :  Zipf
-- Copyright   :  © 2006 Bart Massey
-- License     :  MIT
-- 
-- Maintainer  :  bart@cs.pdx.edu
-- Stability   :  experimental
-- Portability :  portable
-----------------------------------------------------------------------------

import System.Environment
import List
import Char
import qualified Data.Map as Map

-- |Simultaneously lowercase chars and turn all non-alpha
-- chars into whitespace.
clean :: Char -> Char
clean c | isAlpha c = toLower c
clean _ = ' '

-- |A frequency table maps words to counts.
type FTable = Map.Map String Integer

-- |Increment the count of word w in the frequency table m.
increment_word :: FTable -> String -> FTable
increment_word m w = Map.insertWith (+) w 1 m

--- The function foldl is a standard Haskell function for
--- iterating over a list.
---
-- |Given a list of words l, increment the count in the
-- frequency table m for each word in l, returning the
-- resulting table.
count_words :: [String] -> FTable
count_words l = foldl increment_word Map.empty l

-- |Given a word and its frequency, produce the output string in
-- its expected format.
format :: (Integer, Integer) -> String
format (i, c) = show (1.0 / fromIntegral i) ++ " " ++ show c

-- |This convenience operator '>.=' is just the reverse composition
-- operator, here to let us write "pipelines" conveniently.
(>.=) :: (a -> b) -> (b -> c) -> (a -> c)
f >.= g = g . f
infixr 9 >.=

--- The Haskell interact function expects as an argument
--- a function from a list of characters to a list of
--- characters; it applies that function to standard input
--- to produce standard output.  Lazy evaluation makes this
--- a reasonable thing to do.
---
-- |Emit Zipf's Law data for a text presented
--  on standard input.
main :: IO ()
main = do
  interact
    (map clean >.= words >.= count_words >.=
    Map.elems >.= sort >.= reverse >.= zip [1..]
     >.= reverse >.= map format >.= unlines)
